aderyn_core/src/context/browser/chain.rs

```diff
@@ -0,0 +1,580 @@
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for ModifierDefinition {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for ModifierInvocation {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for OverrideSpecifier {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for ParameterList {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for PragmaDirective {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for Return {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for SourceUnit {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for StructDefinition {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for StructuredDocumentation {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for TupleExpression {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for UnaryOperation {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for UserDefinedValueTypeDefinition {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for UsingForDirective {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for VariableDeclaration {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for VariableDeclarationStatement {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+impl GetParentChain for WhileStatement {
+    fn of(&self, context: &WorkspaceContext) -> Vec<ASTNode> {
+        get_parent_chain_of_child(self.id, context)
+    }
+}
+
+pub fn get_parent_chain_of_child(child_id: NodeID, context: &WorkspaceContext) -> Vec<ASTNode> {
```

Comment:
Heart of the logic for this PR

Reply:
This function also serves the next PR https://github.com/Cyfrin/aderyn/pull/167 which allows us to get immediate children

Reply:
My concern with this being here is that this work is done every time a detector calls `GetParent::of`. As the set of detectors grows, the number of calls to this will increase. Obviously, we can introduce caching and other tricks to help the time complexity, but it still feels like an antipattern to do all of this work _each time_.

I wonder if a preprocessing step can prepare the ASTNodes to have a `.parent()` accessor. The preprocessing runs only once when the contract AST files are loaded initially (the entire tree is traversed originally _anyway_). That feels like the more efficient and clean way to do this. Can we piggyback off the initial tree traversal of Aderyn? @TilakMaddy

Reply:
Thanks for the input : ) It definitely feels like an antipattern to do each time and. 

I think what we can do, is like you said preprocess the tree by creating a new hashmap field in workspace context that can give all the IDs of all children nodes for a given node ID.

I don't know if we can piggyback off the first traversals but definitely we can ensure that GetParent on a given node doesn't run twice

Let me try it out. In the meantime if you have any other ideas let me know.